### PR TITLE
Create newsletter form. See #43.

### PR DESCRIPTION
**Pull request by lauracernau**

### CODE REVIEW

File: docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+   *   The subscription form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Storage for the state of the form.
+   * @return \Drupal\Core\Ajax\AjaxResponse $response
+   */
+  public function saveSubscription(array &$form, FormStateInterface $form_state) {
+    $response = new AjaxResponse();
+    $error = FALSE;
+
+    if (empty($form_state->getValue('email'))) {
```
Comment (danielpopdan) [inline docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php]: Please add comments, the logic here is not easily understandable.

File: docroot/modules/custom/dct_newsletter/dct_newsletter.routing.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+dct_newsletter.form:
+  path: '/form'
```
Comment (danielpopdan) [inline docroot/modules/custom/dct_newsletter/dct_newsletter.routing.yml]: The newsletter subscription form should be displayed inside a block.

File: docroot/modules/custom/dct_newsletter/dct_newsletter.info.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+name: DCT Newsletter
+type: module
+description: Provides newsletter functionality.
```
Comment (danielpopdan) [inline docroot/modules/custom/dct_newsletter/dct_newsletter.info.yml]: Please correct to "Provides newsletter related functionality."

File: docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\HtmlCommand;
+use Drupal\Core\Ajax\ReplaceCommand;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Form\FormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Egulias\EmailValidator\EmailValidatorInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+class NewsletterForm extends FormBase {
```
Comment (danielpopdan) [inline docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php]: Please rename to "NewsletterSubscriptionForm".

File: docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        $html = [
+          '#prefix' => '<span class="error-message">',
+          '#markup' => $this->t('This email address is not valid.'),
+          '#suffix' => '</span>',
+        ];
+        $error = TRUE;
+      }
+
+      else {
+        $user_storage = $this->entityTypeManager->getStorage('user');
```
Comment (danielpopdan) [inline docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php]: You can check if the email exists, loading the user by properties. 
$users = \Drupal::entityTypeManager()->getStorage('user')->loadByProperties(['mail' => $form_state->getValue('email')]);

File: docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            '#markup' => $this->t('This email address is already used.'),
+            '#suffix' => '</span>',
+          ];
+          $error = TRUE;
+        }
+      }
+    }
+
+    if ($error == FALSE) {
+
```
Comment (danielpopdan) [inline docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php]: Please remove this empty line. You can if (!$error).

File: docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+      $user = [
+        'name' => $form_state->getValue('email'),
+        'mail' => $form_state->getValue('email'),
+        'init' => $form_state->getValue('email'),
+        'pass' => user_password(),
+        'roles' => ['newsletter'],
+        'status' => 0,
+      ];
+
+      $user = $this->entityTypeManager->getStorage('user')->create($user);
```
Comment (danielpopdan) [inline docroot/modules/custom/dct_newsletter/src/Form/NewsletterForm.php]: Please check $user->isValid() before saving. If something went wrong, set an error message.

File: docroot/modules/custom/dct_newsletter/dct_newsletter.routing.yml (hunk 2 above)
Comment (lauracernau) [reply docroot/modules/custom/dct_newsletter/dct_newsletter.routing.yml]: I added a route just for testing. I will place the form in a block in another sub-task.

**Pull request merged**